### PR TITLE
Ability to read stats from tunnel.

### PR DESCRIPTION
**Pull request by jigar-f**

### CODE REVIEW

**Comment by jigar-f** [comment]
@atavism Can you check if this works for you or not? Also, does this effect start-up time?

**Comment by jigar-f** [comment]
@atavism The only thing is it only checks when the VPN status is connected or connecting if in between server changes then there is no way to update UI. I am working on that...

**Comment by atavism** [comment]
> @atavism The only thing is it only checks when the VPN status is connected or connecting if in between server changes then there is no way to update UI. I am working on that.

@jigar-f I don't see the server change even when I'm connected to the VPN for whatever reason

**Comment by jigar-f** [comment]
Hey @atavism Could you try this now, It should work now, When you start the proxy wait for a few seconds, and you will see data.

**Comment by jigar-f** [comment]
@atavism This should fix all issues related to tunnel to App communication for stats. Even if the server changes app will trigger a different UI.

File: internalsdk/ios/ios.go [hunk 1]
```diff
@@ -190,28 +195,25 @@ func (c *iosClient) start() (ClientWriter, error) {
...
+	if len(dialers) == 0 {
+		return nil, errors.New("No dialers found")
+	}
 	tracker := stats.NewTracker()
 	dialer, err := bandit.NewWithStats(dialers, tracker)
 	if err != nil {
 		return nil, err
 	}
 	go func() {
 		tracker.AddListener(func(st stats.Stats) {
```
Comment (jigar-f) [inline internalsdk/ios/ios.go:213]: @myleshorton question to you

What is supposed to be a time when we create `bandit.NewWithStats` and we will get updated from stats?

Looking at my logs sometimes it takes around 2 min to get a callback

```
May 24 10:43:27.286 - 0m12s DEBUG ios: ios.go:199 User config process start at 2024-05-24 10:43:27.286492 +0000 UTC m=+14.450121584

May 24 10:45:57.004 - 2m42s DEBUG ios: ios.go:215 Stats update at 2024-05-24 10:45:57.003982 +0000 UTC m=+164.167602251
```
cc @atavism

**Comment by jigar-f** [comment]
@atavism works well for me, for the first install it takes a bit of time to show up on the UI, but after that, it works well, Let me know what it goes for you.

**Comment by atavism** [comment]
> @atavism works well for me, for the first install it takes a bit of time to show up on the UI, but after that, it works well, Let me know what it goes for you.

I'll take a look!

**Comment by atavism** [comment]
> Let me know what it goes for you.

Doesn't happen right away, but the server location eventually shows up for me as well!

File: internalsdk/ios/ios.go (hunk 1 above)
Comment (atavism) [reply internalsdk/ios/ios.go:213]: > What is supposed to be the time when we create bandit.NewWithStats and when we will get updated from stats? is there any timeout?

bandit should send updated stats every time there's a successful dial: https://github.com/getlantern/flashlight/blob/main/bandit/bandit.go#L161

File: ios/Runner/Lantern/Models/SessionModel.swift [hunk 2]
```diff
@@ -35,26 +35,93 @@ class SessionModel: BaseModel<InternalsdkSessionModel> {
...
+      } else if let error = error {
+        logger.log("failed to get protoken")
+      }
+    } catch {
+      SentryUtils.caputure(error: error as NSError)
+    }
+  }
+
+  func observeStatsUpdates() {
+    logger.debug("observesing stats udpates")
```
Comment (atavism) [inline ios/Runner/Lantern/Models/SessionModel.swift:92]: Nit: change to "observing stats updates"

File: ios/Runner/Lantern/Models/SessionModel.swift [hunk 3]
```diff
@@ -35,26 +35,93 @@ class SessionModel: BaseModel<InternalsdkSessionModel> {
...
+    do {
+      // Convert the JSON data back to a dictionary
+      if let dataDict = try JSONSerialization.jsonObject(with: stats, options: [])
+        as? [String: Any]
+      {
+        try invoke("updateStats", dataDict)
+        logger.debug("New data received: \(dataDict)")
+      }
+    } catch {
+      logger.debug("Failed to deserialize JSON data: \(error)")
```
Comment (atavism) [inline ios/Runner/Lantern/Models/SessionModel.swift:121]: Should we report any JSON deserialization errors to Sentry as well?

**Review by atavism — APPROVED** [review]
LGTM. Just a couple comments.

File: ios/Runner/Lantern/Models/SessionModel.swift (hunk 2 above)
Comment (jigar-f) [reply ios/Runner/Lantern/Models/SessionModel.swift:92]: Done.

File: ios/Runner/Lantern/Models/SessionModel.swift (hunk 3 above)
Comment (jigar-f) [reply ios/Runner/Lantern/Models/SessionModel.swift:121]: Done.

**Comment by jigar-f** [comment]
> Doesn't happen right away, but the server location eventually shows up for me as well

This depends on when we get an update from the bandit dialer, there is no delay on UI, I tested multiple times. I saw stats getting in under 2 sec, but sometimes it takes some time.

File: ios/Runner/Lantern/Models/SessionModel.swift (hunk 2 above)
Comment (jigar-f) [reply ios/Runner/Lantern/Models/SessionModel.swift:92]: Done.

**Comment by jigar-f** [comment]
@atavism  Are you good with merging to the ios startup branch?

**Comment by atavism** [comment]
@jigar-f I just tested the 7.5.1 release and a server location never shows up for me. I didn't realize some of this stuff was already broken.

**Comment by jigar-f** [comment]
@atavism you mean on iOS production build?

**Pull request merged**